File: shared-contracts/src/bridge/system-state.ts
import { z } from "zod";

import { healthStatusSchema } from "../domain/primitives.js";
import { modeSchema, runtimeStateSchema } from "../domain/runtime.js";
import {
  colorTokenSchema,
  durationSecondsSchema,
  isoUtcSchema,
  localDateSchema,
  opaqueIdSchema,
  progressKindSchema,
  ratioSchema,
  riskLevelSchema,
  schemaVersionSchema,
  sequenceNumberSchema,
  severitySchema,
  uuidSchema,
} from "../domain/scalars.js";

export const explainabilityItemSchema = z.object({
  code: z.string().min(1),
  detail: z.string().min(1),
  weight: z.number(),
});

export const privacyExclusionEntrySchema = z.object({
  exclusion_id: opaqueIdSchema.nullable(),
  label: z.string().min(1),
  match_type: z.enum(["app", "domain", "url_regex", "window_title_regex"]),
  pattern: z.string().min(1),
  enabled: z.boolean(),
});

const bannerViewModelSchema = z.object({
  severity: severitySchema,
  title: z.string().min(1),
  body: z.string().min(1),
});

const plannedTaskViewModelSchema = z.object({
  task_id: opaqueIdSchema,
  title: z.string().min(1),
  success_definition: z.string().min(1),
  total_remaining_effort_seconds: durationSecondsSchema.nullable(),
  intended_work_seconds_today: durationSecondsSchema,
  progress_kind: progressKindSchema,
  allowed_support_work: z.array(z.string()),
  likely_detours_that_still_count: z.array(z.string()),
});

const taskProgressCardSchema = z.object({
  task_id: opaqueIdSchema,
  title: z.string().min(1),
  progress_ratio: ratioSchema.nullable(),
  confidence_ratio: ratioSchema.nullable(),
  risk_level: riskLevelSchema.nullable(),
  aligned_seconds: durationSecondsSchema,
  support_seconds: durationSecondsSchema,
  drift_seconds: durationSecondsSchema,
  eta_remaining_seconds: durationSecondsSchema.nullable(),
  latest_status_text: z.string(),
});

const episodeSummarySchema = z.object({
  episode_id: opaqueIdSchema,
  started_at: isoUtcSchema,
  ended_at: isoUtcSchema,
  runtime_state: runtimeStateSchema,
  matched_task_id: opaqueIdSchema.nullable(),
  matched_task_title: z.string().nullable(),
  is_support_work: z.boolean(),
  confidence_ratio: ratioSchema.nullable(),
  top_evidence: z.array(z.string()),
});

const correctionSummarySchema = z.object({
  correction_id: opaqueIdSchema,
  created_at: isoUtcSchema,
  kind: z.enum(["clarification", "manual_override", "notification_action"]),
  summary_text: z.string().min(1),
});

const ambiguityQueueItemSchema = z.object({
  ambiguity_id: opaqueIdSchema,
  created_at: isoUtcSchema,
  prompt: z.string().min(1),
  status: z.enum(["pending", "resolved", "dismissed"]),
  resolution_summary: z.string().nullable(),
});

const durableRuleReviewItemSchema = z.object({
  review_item_id: opaqueIdSchema,
  created_at: isoUtcSchema,
  title: z.string().min(1),
  rationale: z.string().min(1),
  proposed_rule_text: z.string().min(1),
});

const morningExchangeViewModelSchema = z.object({
  status: z.enum(["required", "available", "completed"]),
  context_packet_text: z.string().nullable(),
  prompt_text: z.string().nullable(),
});

const eveningExchangeViewModelSchema = z.object({
  status: z.enum(["not_ready", "available", "completed"]),
  debrief_packet_text: z.string().nullable(),
  prompt_text: z.string().nullable(),
});

const privacyExclusionsViewModelSchema = z.object({
  exclusions: z.array(privacyExclusionEntrySchema),
});

const dailyPlanViewModelSchema = z.object({
  plan_id: opaqueIdSchema,
  imported_at: isoUtcSchema,
  local_date: localDateSchema,
  total_intended_work_seconds: durationSecondsSchema,
  notes_for_tracker: z.string().nullable(),
  tasks: z.array(plannedTaskViewModelSchema),
});

const dashboardViewModelSchema = z.object({
  header: z.object({
    local_date: localDateSchema,
    mode: modeSchema,
    summary_text: z.string(),
    warning_banner: bannerViewModelSchema.nullable(),
  }),
  plan: dailyPlanViewModelSchema.nullable(),
  current_focus: z.object({
    runtime_state: runtimeStateSchema,
    is_support_work: z.boolean(),
    confidence_ratio: ratioSchema.nullable(),
    explainability: z.array(explainabilityItemSchema),
    last_good_context: z.string().nullable(),
    last_updated_at: isoUtcSchema,
  }),
  progress: z.object({
    total_intended_work_seconds: durationSecondsSchema.nullable(),
    total_aligned_seconds: durationSecondsSchema,
    total_support_seconds: durationSecondsSchema,
    total_drift_seconds: durationSecondsSchema,
    tasks: z.array(taskProgressCardSchema),
  }),
  recent_episodes: z.array(episodeSummarySchema),
  corrections: z.array(correctionSummarySchema),
  ambiguity_queue: z.array(ambiguityQueueItemSchema),
  review_queue: z.array(durableRuleReviewItemSchema),
  morning_exchange: morningExchangeViewModelSchema.nullable(),
  evening_exchange: eveningExchangeViewModelSchema.nullable(),
  privacy_exclusions: privacyExclusionsViewModelSchema,
});

const clarificationChoiceSchema = z.object({
  answer_id: opaqueIdSchema,
  label: z.string().min(1),
  semantics: z.enum([
    "task",
    "support_work",
    "work_group",
    "admin",
    "break",
    "intentional_detour",
    "not_related",
  ]),
  task_id: opaqueIdSchema.nullable(),
  work_group_id: opaqueIdSchema.nullable(),
});

const clarificationHudViewModelSchema = z.object({
  clarification_id: opaqueIdSchema,
  created_at: isoUtcSchema,
  expires_at: isoUtcSchema.nullable(),
  prompt: z.string().min(1),
  subtitle: z.string().nullable(),
  choices: z.array(clarificationChoiceSchema),
  related_episode_id: opaqueIdSchema.nullable(),
  remember_toggle_default: z.boolean(),
  allow_remember_toggle: z.boolean(),
});

const interventionActionSchema = z.object({
  action_id: opaqueIdSchema,
  label: z.string().min(1),
  semantic_action: z.enum([
    "return_now",
    "intentional_detour",
    "pause_10_minutes",
    "open_dashboard",
    "dismiss",
    "confirm_milestone",
    "dismiss_milestone",
  ]),
});

const interventionViewModelSchema = z.object({
  intervention_id: opaqueIdSchema,
  created_at: isoUtcSchema,
  kind: z.enum([
    "hard_drift",
    "praise",
    "recovery_anchor",
    "risk_prompt",
    "clarification_notification",
    "milestone_candidate",
  ]),
  presentation: z.enum(["dashboard_only", "local_notification", "both"]),
  severity: severitySchema,
  title: z.string().min(1),
  body: z.string().min(1),
  actions: z.array(interventionActionSchema),
  suppress_native_notification: z.boolean(),
  suppression_reason: z
    .enum(["observe_only", "cooldown", "paused", "permissions_missing", "mode_gate"])
    .nullable(),
  dedupe_key: z.string().min(1),
  expires_at: isoUtcSchema.nullable(),
});

const menuBarViewModelSchema = z
  .object({
    color_token: colorTokenSchema,
    mode_label: z.string().min(1),
    primary_label: z.string().min(1),
    secondary_label: z.string().nullable(),
    runtime_state: runtimeStateSchema,
    is_support_work: z.boolean(),
    confidence_ratio: ratioSchema.nullable(),
    active_goal_id: opaqueIdSchema.nullable(),
    active_goal_title: z.string().nullable(),
    active_task_id: opaqueIdSchema.nullable(),
    active_task_title: z.string().nullable(),
    state_started_at: isoUtcSchema.nullable(),
    focused_elapsed_seconds: durationSecondsSchema.nullable(),
    pause_until: isoUtcSchema.nullable(),
    allowed_actions: z.object({
      can_pause: z.boolean(),
      can_resume: z.boolean(),
      can_take_break: z.boolean(),
      can_open_morning_flow: z.boolean(),
      can_open_evening_flow: z.boolean(),
    }),
  })
  .superRefine((value, ctx) => {
    if (value.is_support_work && value.runtime_state !== "aligned") {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["is_support_work"],
        message: "Support work is only valid while runtime_state is aligned.",
      });
    }
  });

const systemHealthViewModelSchema = z.object({
  overall_status: healthStatusSchema,
  screenpipe: z.object({
    status: healthStatusSchema,
    last_ok_at: isoUtcSchema.nullable(),
    last_error_at: isoUtcSchema.nullable(),
    message: z.string().nullable(),
  }),
  database: z.object({
    status: healthStatusSchema,
    last_ok_at: isoUtcSchema.nullable(),
    last_error_at: isoUtcSchema.nullable(),
    message: z.string().nullable(),
  }),
  scheduler: z.object({
    fast_tick_last_ran_at: isoUtcSchema.nullable(),
    slow_tick_last_ran_at: isoUtcSchema.nullable(),
  }),
  notifications: z.object({
    os_permission: z.enum(["unknown", "granted", "denied"]),
    muted_by_logic: z.boolean(),
    muted_reason: z
      .enum(["observe_only", "cooldown", "paused", "mode_gate"])
      .nullable(),
  }),
  observe_only: z.object({
    active: z.boolean(),
    ticks_remaining: z.number().int().nonnegative().nullable(),
  }),
});

export const systemStateSchema = z
  .object({
    schema_version: schemaVersionSchema,
    runtime_session_id: uuidSchema,
    stream_sequence: sequenceNumberSchema,
    emitted_at: isoUtcSchema,
    caused_by_command_id: uuidSchema.nullable(),
    mode: modeSchema,
    menu_bar: menuBarViewModelSchema,
    dashboard: dashboardViewModelSchema,
    clarification_hud: clarificationHudViewModelSchema.nullable(),
    intervention: interventionViewModelSchema.nullable(),
    system_health: systemHealthViewModelSchema,
  })
  .superRefine((value, ctx) => {
    if (value.mode === "paused" && value.menu_bar.runtime_state !== "paused") {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["menu_bar", "runtime_state"],
        message: "Paused mode requires menu_bar.runtime_state to be paused.",
      });
    }

    if (value.mode === "no_plan") {
      if (value.menu_bar.active_goal_title !== null) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          path: ["menu_bar", "active_goal_title"],
          message: "No-plan mode requires active goal title to be null.",
        });
      }

      if (value.menu_bar.active_task_title !== null) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          path: ["menu_bar", "active_task_title"],
          message: "No-plan mode requires active task title to be null.",
        });
      }
    }
  });

export type ExplainabilityItem = z.infer<typeof explainabilityItemSchema>;
export type PrivacyExclusionEntry = z.infer<typeof privacyExclusionEntrySchema>;
export type SystemState = z.infer<typeof systemStateSchema>;

File: INeedABossAgent/App/Bridge/SystemState.swift
import Foundation

enum BridgeJSONCoding {
  static let decoder: JSONDecoder = {
    let decoder = JSONDecoder()
    decoder.keyDecodingStrategy = .convertFromSnakeCase
    return decoder
  }()

  static let encoder: JSONEncoder = {
    let encoder = JSONEncoder()
    encoder.keyEncodingStrategy = .convertToSnakeCase
    encoder.outputFormatting = [.sortedKeys]
    return encoder
  }()
}

typealias SchemaVersion = String

enum Mode: String, Codable, Equatable {
  case booting
  case noPlan = "no_plan"
  case running
  case paused
  case degradedScreenpipe = "degraded_screenpipe"
  case logicError = "logic_error"
}

enum RuntimeState: String, Codable, Equatable {
  case aligned
  case uncertain
  case softDrift = "soft_drift"
  case hardDrift = "hard_drift"
  case paused
}

enum Severity: String, Codable, Equatable {
  case info
  case warning
  case critical
}

enum ColorToken: String, Codable, Equatable {
  case green
  case blue
  case yellow
  case red
  case gray
}

enum RiskLevel: String, Codable, Equatable {
  case low
  case medium
  case high
}

enum ProgressKind: String, Codable, Equatable {
  case timeBased = "time_based"
  case milestoneBased = "milestone_based"
  case artifactBased = "artifact_based"
  case hybrid
}

enum CorrectionKind: String, Codable, Equatable {
  case clarification
  case manualOverride = "manual_override"
  case notificationAction = "notification_action"
}

enum AmbiguityStatus: String, Codable, Equatable {
  case pending
  case resolved
  case dismissed
}

enum MorningExchangeStatus: String, Codable, Equatable {
  case requiredStatus = "required"
  case available
  case completed
}

enum EveningExchangeStatus: String, Codable, Equatable {
  case notReady = "not_ready"
  case available
  case completed
}

enum PrivacyMatchType: String, Codable, Equatable {
  case app
  case domain
  case urlRegex = "url_regex"
  case windowTitleRegex = "window_title_regex"
}

enum ClarificationChoiceSemantic: String, Codable, Equatable {
  case task
  case supportWork = "support_work"
  case workGroup = "work_group"
  case admin
  case `break`
  case intentionalDetour = "intentional_detour"
  case notRelated = "not_related"
}

enum InterventionKind: String, Codable, Equatable {
  case hardDrift = "hard_drift"
  case praise
  case recoveryAnchor = "recovery_anchor"
  case riskPrompt = "risk_prompt"
  case clarificationNotification = "clarification_notification"
}

enum InterventionPresentation: String, Codable, Equatable {
  case dashboardOnly = "dashboard_only"
  case localNotification = "local_notification"
  case both
}

enum InterventionSemanticAction: String, Codable, Equatable {
  case returnNow = "return_now"
  case intentionalDetour = "intentional_detour"
  case pause10Minutes = "pause_10_minutes"
  case openDashboard = "open_dashboard"
  case dismiss
}

enum InterventionSuppressionReason: String, Codable, Equatable {
  case observeOnly = "observe_only"
  case cooldown
  case paused
  case permissionsMissing = "permissions_missing"
  case modeGate = "mode_gate"
}

enum HealthStatus: String, Codable, Equatable {
  case ok
  case degraded
  case down
}

enum NotificationPermissionStatus: String, Codable, Equatable {
  case unknown
  case granted
  case denied
}

enum NotificationMuteReason: String, Codable, Equatable {
  case observeOnly = "observe_only"
  case cooldown
  case paused
  case modeGate = "mode_gate"
}

struct SystemState: Codable, Equatable {
  let schemaVersion: SchemaVersion
  let runtimeSessionId: String
  let streamSequence: Int
  let emittedAt: String
  let causedByCommandId: String?
  let mode: Mode
  let menuBar: MenuBarViewModel
  let dashboard: DashboardViewModel
  let clarificationHud: ClarificationHudViewModel?
  let intervention: InterventionViewModel?
  let systemHealth: SystemHealthViewModel
}

struct MenuBarViewModel: Codable, Equatable {
  let colorToken: ColorToken
  let modeLabel: String
  let primaryLabel: String
  let secondaryLabel: String?
  let runtimeState: RuntimeState
  let isSupportWork: Bool
  let confidenceRatio: Double?
  let activeGoalId: String?
  let activeGoalTitle: String?
  let activeTaskId: String?
  let activeTaskTitle: String?
  let stateStartedAt: String?
  let focusedElapsedSeconds: Int?
  let pauseUntil: String?
  let allowedActions: AllowedActionsViewModel
}

struct AllowedActionsViewModel: Codable, Equatable {
  let canPause: Bool
  let canResume: Bool
  let canTakeBreak: Bool
  let canOpenMorningFlow: Bool
  let canOpenEveningFlow: Bool
}

struct DashboardViewModel: Codable, Equatable {
  let header: DashboardHeaderViewModel
  let plan: DailyPlanViewModel?
  let currentFocus: CurrentFocusViewModel
  let progress: ProgressSummaryViewModel
  let recentEpisodes: [EpisodeSummary]
  let corrections: [CorrectionSummary]
  let ambiguityQueue: [AmbiguityQueueItem]
  let reviewQueue: [DurableRuleReviewItem]
  let morningExchange: MorningExchangeViewModel?
  let eveningExchange: EveningExchangeViewModel?
  let privacyExclusions: PrivacyExclusionsViewModel
}

struct DashboardHeaderViewModel: Codable, Equatable {
  let localDate: String
  let mode: Mode
  let summaryText: String
  let warningBanner: BannerViewModel?
}

struct BannerViewModel: Codable, Equatable {
  let severity: Severity
  let title: String
  let body: String
}

struct DailyPlanViewModel: Codable, Equatable {
  let planId: String
  let importedAt: String
  let localDate: String
  let totalIntendedWorkSeconds: Int
  let notesForTracker: String?
  let tasks: [PlannedTaskViewModel]
}

struct PlannedTaskViewModel: Codable, Equatable {
  let taskId: String
  let title: String
  let successDefinition: String
  let totalRemainingEffortSeconds: Int?
  let intendedWorkSecondsToday: Int
  let progressKind: ProgressKind
  let allowedSupportWork: [String]
  let likelyDetoursThatStillCount: [String]
}

struct CurrentFocusViewModel: Codable, Equatable {
  let runtimeState: RuntimeState
  let isSupportWork: Bool
  let confidenceRatio: Double?
  let explainability: [ExplainabilityItem]
  let lastGoodContext: String?
  let lastUpdatedAt: String
}

struct ExplainabilityItem: Codable, Equatable {
  let code: String
  let detail: String
  let weight: Double
}

struct ProgressSummaryViewModel: Codable, Equatable {
  let totalIntendedWorkSeconds: Int?
  let totalAlignedSeconds: Int
  let totalSupportSeconds: Int
  let totalDriftSeconds: Int
  let tasks: [TaskProgressCard]
}

struct TaskProgressCard: Codable, Equatable {
  let taskId: String
  let title: String
  let progressRatio: Double?
  let confidenceRatio: Double?
  let riskLevel: RiskLevel?
  let alignedSeconds: Int
  let supportSeconds: Int
  let driftSeconds: Int
  let etaRemainingSeconds: Int?
  let latestStatusText: String
}

struct EpisodeSummary: Codable, Equatable {
  let episodeId: String
  let startedAt: String
  let endedAt: String
  let runtimeState: RuntimeState
  let matchedTaskId: String?
  let matchedTaskTitle: String?
  let isSupportWork: Bool
  let confidenceRatio: Double?
  let topEvidence: [String]
}

struct CorrectionSummary: Codable, Equatable {
  let correctionId: String
  let createdAt: String
  let kind: CorrectionKind
  let summaryText: String
}

struct AmbiguityQueueItem: Codable, Equatable {
  let ambiguityId: String
  let createdAt: String
  let prompt: String
  let status: AmbiguityStatus
  let resolutionSummary: String?
}

struct DurableRuleReviewItem: Codable, Equatable {
  let reviewItemId: String
  let createdAt: String
  let title: String
  let rationale: String
  let proposedRuleText: String
}

struct MorningExchangeViewModel: Codable, Equatable {
  let status: MorningExchangeStatus
  let contextPacketText: String?
  let promptText: String?
}

struct EveningExchangeViewModel: Codable, Equatable {
  let status: EveningExchangeStatus
  let debriefPacketText: String?
  let promptText: String?
}

struct PrivacyExclusionsViewModel: Codable, Equatable {
  let exclusions: [PrivacyExclusionEntry]
}

struct PrivacyExclusionEntry: Codable, Equatable {
  let exclusionId: String?
  let label: String
  let matchType: PrivacyMatchType
  let pattern: String
  let enabled: Bool
}

struct ClarificationHudViewModel: Codable, Equatable {
  let clarificationId: String
  let createdAt: String
  let expiresAt: String?
  let prompt: String
  let subtitle: String?
  let choices: [ClarificationChoice]
  let relatedEpisodeId: String?
  let rememberToggleDefault: Bool
  let allowRememberToggle: Bool
}

struct ClarificationChoice: Codable, Equatable {
  let answerId: String
  let label: String
  let semantics: ClarificationChoiceSemantic
  let taskId: String?
  let workGroupId: String?
}

struct InterventionViewModel: Codable, Equatable {
  let interventionId: String
  let createdAt: String
  let kind: InterventionKind
  let presentation: InterventionPresentation
  let severity: Severity
  let title: String
  let body: String
  let actions: [InterventionAction]
  let suppressNativeNotification: Bool
  let suppressionReason: InterventionSuppressionReason?
  let dedupeKey: String
  let expiresAt: String?
}

struct InterventionAction: Codable, Equatable {
  let actionId: String
  let label: String
  let semanticAction: InterventionSemanticAction
}

struct SystemHealthViewModel: Codable, Equatable {
  let overallStatus: HealthStatus
  let screenpipe: SystemComponentHealth
  let database: SystemComponentHealth
  let scheduler: SchedulerHealth
  let notifications: NotificationHealth
  let observeOnly: ObserveOnlyHealth
}

struct SystemComponentHealth: Codable, Equatable {
  let status: HealthStatus
  let lastOkAt: String?
  let lastErrorAt: String?
  let message: String?
}

struct SchedulerHealth: Codable, Equatable {
  let fastTickLastRanAt: String?
  let slowTickLastRanAt: String?
}

struct NotificationHealth: Codable, Equatable {
  let osPermission: NotificationPermissionStatus
  let mutedByLogic: Bool
  let mutedReason: NotificationMuteReason?
}

struct ObserveOnlyHealth: Codable, Equatable {
  let active: Bool
  let ticksRemaining: Int?
}
